google/cloud/spanner_v1/session.py

```diff
@@ -458,47 +463,100 @@ def run_in_transaction(self, func, *args, **kw):
-                txn = self.transaction()
-                txn.transaction_tag = transaction_tag
-                txn.exclude_txn_from_change_streams = exclude_txn_from_change_streams
-            else:
-                txn = self._transaction
+        observability_options = getattr(self._database, "observability_options", None)
+        with trace_call(
+            "CloudSpanner.Session.run_in_transaction",
+            self,
+            observability_options=observability_options,
+        ) as span:
+            while True:
+                if self._transaction is None:
+                    add_span_event(span, "Creating Transaction")
+                    txn = self.transaction()
+                    txn.transaction_tag = transaction_tag
+                    txn.exclude_txn_from_change_streams = (
+                        exclude_txn_from_change_streams
+                    )
+                else:
+                    txn = self._transaction
 
-            try:
                 attempts += 1
-                return_value = func(txn, *args, **kw)
-            except Aborted as exc:
-                del self._transaction
-                _delay_until_retry(exc, deadline, attempts)
-                continue
-            except GoogleAPICallError:
-                del self._transaction
-                raise
-            except Exception:
-                txn.rollback()
-                raise
-
-            try:
-                txn.commit(
-                    return_commit_stats=self._database.log_commit_stats,
-                    request_options=commit_request_options,
-                    max_commit_delay=max_commit_delay,
-                )
-            except Aborted as exc:
-                del self._transaction
-                _delay_until_retry(exc, deadline, attempts)
-            except GoogleAPICallError:
-                del self._transaction
-                raise
-            else:
-                if self._database.log_commit_stats and txn.commit_stats:
-                    self._database.logger.info(
-                        "CommitStats: {}".format(txn.commit_stats),
-                        extra={"commit_stats": txn.commit_stats},
+                span_attributes = {"attempt": attempts}
+                txn_id = getattr(txn, "_transaction_id", "") or ""
+                if txn_id:
+                    span_attributes["transaction.id"] = txn_id
+
+                add_span_event(span, "Using Transaction", span_attributes)
+
+                try:
+                    return_value = func(txn, *args, **kw)
+                except Aborted as exc:
+                    del self._transaction
+                    if span:
+                        delay_seconds = _get_retry_delay(exc.errors[0], attempts)
+                        attributes = dict(delay_seconds=delay_seconds)
+                        attributes.update(span_attributes)
+                        record_span_exception_and_status(span, exc)
```

Comment:
I think we don't need these at all?
Here lets say there are x retries then the retry_delay getting captured will be overwritten by the xth retry value and the retry delayas of 1 to x-1 retries will not be captued. I don't see any benefit of this.

Reply:
It won't be overwritten, adding a span event creates a fresh one even if with the exact same event_name and attributes because events are recorded in time and not as a lookup; it is useful to capture the delays and retries and record the events because then customers do not have to guess about what happened where, they'll clearly see the report of a retry attempt and with how much delay.